### PR TITLE
build(mge): clang do not support compile option: -fuse-ld=gold

### PR DESCRIPTION
**Pull request by KumaAnjo**
[  1%] Building CXX object flatbuffers/CMakeFiles/flatc.dir/src/idl_gen_text.cpp.o
[  1%] Performing configure step for 'mkl_dnn'
clang: error: argument unused during compilation: '-fuse-ld=gold' [-Werror,-Wunused-command-line-argument]
flatbuffers/CMakeFiles/flatc.dir/build.make:86: recipe for target 'flatbuffers/CMakeFiles/flatc.dir/src/idl_gen_text.cpp.o' failed
make[2]: *** [flatbuffers/CMakeFiles/flatc.dir/src/idl_gen_text.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
clang: error: argument unused during compilation: '-fuse-ld=gold' [-Werror,-Wunused-command-line-argument]
flatbuffers/CMakeFiles/flatc.dir/build.make:62: recipe for target 'flatbuffers/CMakeFiles/flatc.dir/src/idl_parser.cpp.o' failed
make[2]: *** [flatbuffers/CMakeFiles/flatc.dir/src/idl_parser.cpp.o] Error 1
CMakeFiles/Makefile2:379: recipe for target 'flatbuffers/CMakeFiles/flatc.dir/all' failed
make[1]: *** [flatbuffers/CMakeFiles/flatc.dir/all] Error 2

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -36,8 +36,10 @@ endif()
 CHECK_CXX_COMPILER_FLAG(-fuse-ld=gold CXX_SUPPORT_GOLD)
 if(CXX_SUPPORT_GOLD)
     message("-- Using GNU gold linker.")
-    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -fuse-ld=gold")
-    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fuse-ld=gold")
+    if (CMAKE_CXX_COMPILER_ID MATCHES "GNU")
+        set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -fuse-ld=gold")
+        set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fuse-ld=gold")
+    endif()
```
Comment (tosytr) [inline CMakeLists.txt:42]:
```suggestion
    set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_C_FLAGS} -fuse-ld=gold")
    set(CMAKE_SHARED_LINKER_FLAGS "${CMAKE_C_FLAGS} -fuse-ld=gold")
```

Comment (KumaAnjo) [reply CMakeLists.txt:42]: clang do not support compile options, but support link options.
ok, i will re-commit

**Comment by KumaAnjo** [comment]
ci snippet:
2020-04-21T06:50:16.4586206Z [ RUN      ] TestAsyncWorker.FutureThreadPoolException
2020-04-21T06:50:16.6828195Z /__w/MegEngine/MegEngine/src/core/test/utils/async_worker.cpp:147: Failure
2020-04-21T06:50:16.6828455Z Expected: (time) < (0.21), actual: 0.211967 vs 0.21
2020-04-21T06:50:16.6828705Z [  FAILED  ] TestAsyncWorker.FutureThreadPoolException (224 ms)
2020-04-21T06:50:16.6829874Z [----------] 4 tests from TestAsyncWorker (705 ms total)

CI test fail when only modify CMakeLists.txt.

**Comment by yaochengji** [comment]
Thanks for your update, @KumaAnjo . I think you could just replace the two lines with
```
set(CMAKE_SHARED_LINKER_FLAGS "${CMAKE_SHARED_LINKER_FLAGS} ${MGE_COMMON_LD}")
set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} ${MGE_COMMON_LD}")
```
`if else` is not needed.

**Comment by tosytr** [comment]
> https://pipelines.actions.githubusercontent.com/h1fZzzIpWomZyijGy5zMEqJyhLivUM0lTjVxmX7dRFxqhDu8sS/_apis/pipelines/1/runs/13/signedlogcontent/3?urlExpires=2020-04-21T08%3A29%3A16.1223228Z&urlSigningMethod=HMACV1&urlSignature=pWlsWWVwhxxLQOHLcoyP%2BA%2F40%2B1uIesjgUzuMxwf6mg%3D
> 
> ci snippet:
> 2020-04-21T06:50:16.4586206Z [ RUN ] TestAsyncWorker.FutureThreadPoolException
> 2020-04-21T06:50:16.6828195Z /__w/MegEngine/MegEngine/src/core/test/utils/async_worker.cpp:147: Failure
> 2020-04-21T06:50:16.6828455Z Expected: (time) < (0.21), actual: 0.211967 vs 0.21
> 2020-04-21T06:50:16.6828705Z [ FAILED ] TestAsyncWorker.FutureThreadPoolException (224 ms)
> 2020-04-21T06:50:16.6829874Z [----------] 4 tests from TestAsyncWorker (705 ms total)
> 
> CI test fail when only modify CMakeLists.txt.

Simply ignore it, it's a flaky test unrelated to this change.

**Comment by xxr3376** [comment]
@KumaAnjo Would you please rebase master and resolve comments?

**Comment by ChaiMind** [comment]
[This branch](https://github.com/KumaAnjo/MegEngine) is 5 commits ahead, 283 commits behind MegEngine:master,
which means you may need handle with lots of conflicts.

So I suggest @KumaAnjo just push a new pull request following https://github.com/MegEngine/MegEngine/pull/44#issuecomment-617084890 . 

> You should produce bitcode files from clang with the option -flto. This flag will also cause clang to look for the gold plugin in the lib directory under its prefix and pass the -plugin option to ld. It will not look for an alternate linker without -fuse-ld=gold, which is why you otherwise need gold to be the installed system linker in your path.

File `src/CMakeLists.txt` is updated now so we can remain it.

Refs:
- [The LLVM gold plugin](https://llvm.org/docs/GoldPlugin.html)
- [Clang++ and `-fuse-ld=gold` results in many unused option warnings](https://stackoverflow.com/questions/43632700/clang-and-fuse-ld-gold-results-in-many-unused-option-warnings)

**Comment by ChaiMind** [comment]
Hey @KumaAnjo ,  I have updated the gold linker  flag setting in #74 .

If you have any other questions, please let me know. 😄 :

**Pull request closed without merging**